### PR TITLE
feat: `CDK::GetProp`

### LINKED ISSUE #245: Use `Fn::GetProp` to get values from CDK constructs
`{ Fn::GetProp }` is the deCDK counterpart to `{ Fn::GetAtt }`. The result is always a deCDK value.

 `{CDK::GetProp: [x, 'Prop']}` | Used in CFN context | Used in deCDK context
-- | -- | --
x = CfnResource | (error) | (error)
x = CdkConstruct | assertCfnValue(x[att]) | x[att]

`{ Fn::GetAtt }` will always retrieve CloudFormation attributes (for reasons below). Therefore, the result is always a CFN value.

 `{Fn::GetAtt: [x, 'CfnAtt']}`  | Used in CFN context | Used in deCDK context
-- | -- | --
x = CfnResource  | x[att] | x[att]
x = CdkConstruct | findDefaultCfnResource(x).getAtt(att)  | findDefaultCfnResource(x).getAtt(att) 

---
## Why `{ Fn::GetAtt }` only references CloudFormation attributes

In the current deCDK, `{ Fn::GetAtt }` allows referencing properties on L2s. For example:
```
Resources:
  CfnResource: # Some CFN resource
  DecdkResource: # Some deCDK resource
  
  OtherResource: 
    Properties:
      Something: { 'Fn::GetAtt': ['DecdkResource', 'decdkProp'] } // -> decdkResource.decdkProp
```
If we continue to allow this, there will be an inconsistency with `{ Ref }:`

* `{ Ref }` will have its semantics consistent with CFN `{ Ref }` , regardless of whether you are referencing a CFN value or a deCDK value.
* `{ Fn::GetAtt } `behaves differently depending on whether you reference a CFN value or a deCDK value:
```
Resources:
  CfnResource: # Some CFN resource
  DecdkResource: # Some deCDK resource
  
  OtherResource:
    Properties:
      Something1: { 'Ref': CfnResource }    # Works
      Something2: { 'Ref': DecdkResource }  # Works
      Something3: { 'Fn::GetAtt': ['CfnResource', 'CfnAttr'] }   # Works
      Something4: { 'Fn::GetAtt': ['DecdkResource', 'CfnAttr'] } # DOES NOT WORK
```
This behavior is confusing, and you might even need a second to figure out why: `CfnAttr` does not exist on the CDK construct (it's a CFN attribute). Even worse when the CFN attr and CDK prop are named the same: `Name` vs `name`. The difference becomes only the case of the first letter. Which is why we make a distinction between `{ Fn::GetAtt }` and `{ Fn::GetProp }`.

### PR DESCRIPTION
**Pull request by mrgrain**
Fixes #245 #268

--- 

Contains several changes:

- Refactor Json Schema for Intrinsics to be more dynamic
- Add `CDK::GetProp` JSON schema
- Refactor stack references to be more generic and support different types
- Change `Fn::GetAtt` to always operate on CFN Resources and always return a intrinsic function
- Add `CDK::GetProp` to get Object properties

### CODE REVIEW

File: src/evaluate/evaluate.ts [hunk 1]
```diff
@@ -175,8 +191,13 @@ export class Evaluator {
     method: string,
     parameters: any[]
   ) {
-    const record = this.context.referenceable(logicalId);
-    const construct = record.primaryValue as any;
+    const record = this.context.reference(logicalId);
+    const construct = record.instance as any;
+    if (!construct || !construct?.[method]) {
```
Comment (otaviomacedo) [inline src/evaluate/evaluate.ts:196]: Did you find a case where this constraint was violated? It should have been caught at type resolution.

File: src/evaluate/references.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+export class InstanceReference extends SimpleReference {
+  public constructor(logicalId: string, public readonly instance: any) {
+    super(logicalId);
+  }
+}
+
+export class ValueOnlyReference extends InstanceReference {
+  public get ref(): string {
+    throw new Error(`Ref: ${this.logicalId} cannot be referenced`);
```
Comment (otaviomacedo) [inline src/evaluate/references.ts:53]: Is this necessarily true? A lazy resource can return a construct (e.g., `aws_lambda.Function.addAlias`).

File: src/evaluate/evaluate.ts (hunk 1 above)
Comment (mrgrain) [reply src/evaluate/evaluate.ts:196]: Good point. I think I had added it for type narrowing. Will take another look.

Comment (mrgrain) [reply src/evaluate/evaluate.ts:196]: Simplified the code without the check.

File: src/evaluate/references.ts (hunk 2 above)
Comment (mrgrain) [reply src/evaluate/references.ts:53]: Fixed.

**Review by otaviomacedo — APPROVED** [review]

**Pull request merged**